src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/EnumTypeModelBinder.cs

```diff
@@ -0,0 +1,121 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel;
+using System.Runtime.ExceptionServices;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Mvc.ModelBinding.Binders
+{
+    /// <summary>
+    /// <see cref="IModelBinder"/> implementation to bind models for types deriving from <see cref="System.Enum"/>.
+    /// </summary>
+    public class EnumTypeModelBinder : IModelBinder
+    {
+        private readonly MvcOptions _mvcOptions;
+
+        public EnumTypeModelBinder(MvcOptions mvcOptions)
+        {
+            _mvcOptions = mvcOptions;
+        }
+
+        /// <inheritdoc />
+        public Task BindModelAsync(ModelBindingContext bindingContext)
+        {
+            if (bindingContext == null)
+            {
+                throw new ArgumentNullException(nameof(bindingContext));
+            }
+
+            var valueProviderResult = bindingContext.ValueProvider.GetValue(bindingContext.ModelName);
+            if (valueProviderResult == ValueProviderResult.None)
+            {
+                // no entry
+                return Task.CompletedTask;
+            }
+
+            bindingContext.ModelState.SetModelValue(bindingContext.ModelName, valueProviderResult);
+            var modelType = bindingContext.ModelMetadata.UnderlyingOrModelType;
+            var stringValue = valueProviderResult.FirstValue;
+
+            try
+            {
+                object model;
+                if (string.IsNullOrWhiteSpace(stringValue))
+                {
+                    model = null;
+                }
+                else
+                {
+                    var typeConverter = TypeDescriptor.GetConverter(modelType);
+                    model = typeConverter.ConvertFrom(
+                        context: null,
+                        culture: valueProviderResult.Culture,
+                        value: stringValue);
+                }
+
+                // Check if the converted value is indeed defined on the enum as EnumConverter converts 
+                // value to the backing type (ex: integer) and does not check if the value is defined on the enum.
+                if (!_mvcOptions.AllowBindingUndefinedValueToEnumType && model != null)
+                {
+                    var isFlagsEnum = modelType.IsDefined(typeof(FlagsAttribute), inherit: false);
```

Comment:
You should cache this kind of detail in the constructor.

Reply:
Information is already cached in `ModelMetadata.IsFlagsEnum`.